arangod/Cluster/SynchronizeShard.cpp

```diff
@@ -384,89 +384,117 @@ static inline bool isStopping() {
+  // This is done automatically below. 
 
-  double sleepTime = 0.5;
-  size_t count = 0;
-  size_t maxTries = static_cast<size_t>(std::floor(600.0 / sleepTime));
-  while (++count < maxTries) {  // wait for some time until read lock established:
+  while (true) {  // wait for some time until read lock established:
 
+    // Exit immediately on shutdown 
     if (isStopping()) {
       return arangodb::Result(TRI_ERROR_SHUTTING_DOWN);
     }
 
-    // Now check that we hold the read lock:
-    auto putres = cc->syncRequest(clientId, 1, endpoint, rest::RequestType::PUT,
-                                  url, body.toJson(),
-                                  std::unordered_map<std::string, std::string>(), timeout);
+    // Timeout. Just to make sure that we are not holding the lock accidentally,
+    // we cannot exit the function here.
+    if (duration_cast<seconds>(steady_clock::now() - start).count() > timeout) {
+      break;
+    }
 
-    auto result = putres->result;
-    if (result != nullptr && result->getHttpReturnCode() == 200) {
-      auto const vp = putres->result->getBodyVelocyPack();
-      auto const& slice = vp->slice();
-      TRI_ASSERT(slice.isObject());
-      if (slice.hasKey("lockHeld") && slice.get("lockHeld").isBoolean() &&
-          slice.get("lockHeld").getBool()) {
+    // See if above POST has been returned (>= 3.4 only)
+    auto enqres = cc->enquire(postres);
+    if (enqres.status == CL_COMM_SENT) {
+      if (enqres.result != nullptr && enqres.result->getHttpReturnCode() == 200) {
+        // Habemus clausum 
         return arangodb::Result();
       }
-      LOG_TOPIC(DEBUG, Logger::MAINTENANCE)
-          << "startReadLockOnLeader: Lock not yet acquired...";
-    } else {
-      LOG_TOPIC(DEBUG, Logger::MAINTENANCE)
-          << "startReadLockOnLeader: Do not see read lock yet:"
-          << putres->stringifyErrorMessage();
     }
 
-    std::this_thread::sleep_for(duration<double>(sleepTime));
+    // Now check that we have acquired the read lock:
+    auto putres = cc->syncRequest(
+      clientId, 1, endpoint, rest::RequestType::PUT, url, body.toJson(),
+      std::unordered_map<std::string, std::string>(), 15.0);
+
+    auto result = putres->result;
+    if (result != nullptr) {
+      if (result->getHttpReturnCode() == 200) {
+        auto const vp = putres->result->getBodyVelocyPack();
+        auto const& slice = vp->slice();
+        TRI_ASSERT(slice.isObject());
+        if (slice.hasKey("lockHeld") && slice.get("lockHeld").isBoolean() &&
+            slice.get("lockHeld").getBool()) {    // Habemus clausum
+          return arangodb::Result();
+        }
+        LOG_TOPIC(DEBUG, Logger::MAINTENANCE)
+          << "startReadLockOnLeader: Lock not yet acquired, retrying... ";
+      } else if (enqres.status == CL_COMM_SENT) { // POST sent in meantime: Error (>=3.4)
+        LOG_TOPIC(DEBUG, Logger::MAINTENANCE)
+          << "startReadLockOnLeader: Failed to acquire read lock: "
```

Comment:
Should we not bail out and give up here? Further waiting does not make sense.

Reply:
done.